packages/app/src/runner/unifiedRunner.ts

```diff
@@ -20,45 +19,39 @@ export function useUnifiedRunner () {
     initialized.value = false
   })
 
-  return {
-    initialized: readonly(initialized),
-
-    watchSpec: (specs: Ref<ReadonlyArray<SpecFile>>) => {
-      const specStore = useSpecStore()
-      const route = useRoute()
-      const selectorPlaygroundStore = useSelectorPlaygroundStore()
+  const specStore = useSpecStore()
+  const route = useRoute()
+  const selectorPlaygroundStore = useSelectorPlaygroundStore()
 
-      watch(() => specs.value, (newVal) => {
-        const fileParam = getPathForPlatform(route.query.file as string)
+  watchEffect(() => {
+    const queryFile = getPathForPlatform(route.query.file as string)
 
-        if (!fileParam) {
-          // no file param, we are not showing a file
-          // so no action needed when specs list updates
-          return
-        }
+    if (!queryFile) {
+      // no file param, we are not showing a file
+      // so no action needed when specs list updates
+      return
+    }
 
-        const activeSpecInSpecsList = newVal.find((x) => x.relative === fileParam)
+    const activeSpecInSpecsList = specs.value.find((x) => x.relative === queryFile)
 
-        if (!activeSpecInSpecsList) {
-          // the specs list no longer contains the spec being shown
-          // so set active state to null and let the UI handle it
-          specStore.setActiveSpec(null)
-        }
-      })
-
-      return watch(() => getPathForPlatform(route.query.file as string), (queryParam) => {
-        const spec = specs.value.find((x) => x.relative === queryParam)
+    if (activeSpecInSpecsList && specStore.activeSpec?.relative !== activeSpecInSpecsList.relative) {
+      specStore.setActiveSpec(activeSpecInSpecsList)
+    } else if (!activeSpecInSpecsList) {
+      specStore.setActiveSpec(null)
+    }
+  })
 
-        if (selectorPlaygroundStore.show) {
-          const autIframe = getAutIframeModel()
+  watch(() => getPathForPlatform(route.query.file as string), (newQueryFile) => {
```

Comment:
We have two things watching `route.query.file`, line 26 and line 44. I wonder if we should combine them so all the watching of the same variable is in the same function? There's an argument of "one watch per effect" so again I don't mind either way, but wanted to point this out since you might not have noticed the double watcher.

I'm guessing the perf overhead is negligible so just wanted to call this out to see if it was intentional or not.

I tried the single watcher and it seems to work just fine, too: https://github.com/cypress-io/cypress/pull/21112

Reply:
Yeah I played around with combining them at one point and ran into what @marktnoonan is talking about. Specifically I think it's these tests failing on #21112: https://app.circleci.com/pipelines/github/cypress-io/cypress/36466/workflows/828e7886-b042-41de-8753-e693c6f9a292/jobs/1460784

There may be a way to tweak the logic to fix that, but ultimately, I'm ok with keeping them separate since I do like the idea of "one watch per effect"

Reply:
"one watch per effect" seems like a reasonable approach to take until it becomes a performance issue. I prefer the sensible encapsulation over worrying about the lifecycle too much.